pkg/vsphere/vm/vm.go

```diff
@@ -712,3 +714,133 @@ func (vm *VirtualMachine) RemoveSnapshotByRef(ctx context.Context, snapshot *typ
 
 	return object.NewTask(vm.Vim25(), res.Returnval), nil
 }
+
+// CheckHostCompatibility checks that a host is compatible for VM placement.
+func (vm *VirtualMachine) CheckHostCompatibility(ctx context.Context, host *types.ManagedObjectReference) (*object.Task, error) {
+	req := types.CheckCompatibility_Task{
+		This: vm.ServiceContent.VmCompatibilityChecker.Reference(),
+		Vm:   vm.Reference(),
+		Host: host,
+	}
+
+	res, err := methods.CheckCompatibility_Task(ctx, vm.Vim25(), &req)
+	if err != nil {
+		return nil, err
+	}
+	return object.NewTask(vm.Vim25(), res.Returnval), nil
+}
+
+// PowerOn powers on a VM. If the environment is VC without DRS enabled, it will attempt to relocate  the VM
+// to the most suitable host in the cluster. If relocation or subsequent power-on fail, it will attempt the next
+// best host, and repeat this process until a successful power-on is achieved or there are no more hosts to try.
+func (vm *VirtualMachine) PowerOn(op trace.Operation) error {
+	drs := vm.Session.DRSEnabled != nil && *vm.Session.DRSEnabled
+	cls := vm.InCluster(op)
+
+	op.Debugf("DRS: %t", drs)
+	op.Debugf("In cluster: %t", cls)
+
+	// any of the following conditions are sufficient to power-on the VM:
+	// - environment is not VC
+	// - DRS is enabled
+	// - the VM does not belong to a cluster
+	// - the current host is adequate for power-on
+	if !vm.IsVC() || drs || !cls || rhp.CheckHost(op, vm.VirtualMachine) {
+		return vm.powerOn(op)
+	}
+
+	// otherwise place the VM before powering it on
+	hmp := performance.NewHostMetricsProvider(vm.Session.Vim25())
+	rhp, err := placement.NewRankedHostPolicy(op, vm.Cluster, hmp)
+	if err != nil {
+		return err
+	}
+
+	op.Warnf("Host is not adequate for power-on, getting placement recommendation")
+
+	hosts, err := rhp.RecommendHost(op, nil)
+	if err != nil {
+		return err
+	}
+
+	var subset []*object.HostSystem
+
+	for err == nil && len(hosts) > 0 {
+		op.Infof("Placement recommended %s", hosts[0].Reference().String())
+
+		if vm.relocate(op, hosts[0]) == nil && vm.powerOn(op) == nil {
+			return nil
+		}
+
+		// if relocation or powerOn fails, something has changed and we need a new recommendation
+		subset = hosts[1:]
+
+		hosts, err = rhp.RecommendHost(op, subset)
+		if err != nil {
```

Comment:
If we have an error the loop will exit -- so why include the error eval in the loop and why not return with specific error here?  If we are not retrying the recommendation then this needs to be `Errorf` and not `Warnf`..

Also why not retry the recommendation?  Is that not a valid scenario?